File: dart/plugins/azuresecrets.dart
// Azure Key Vault, as a voxgig/plugin definition.
//
// A PLUGIN, not a built-in: it opens a socket. A `Sekreto` can build the
// `azuresecrets` kind only if the calling project imported this file and
// passed `azuresecrets` in the `plugins` option
// (docs/design/plugin-providers.md).
//
// A port of typescript/plugins/azuresecrets.ts, which is canonical.

import 'dart:async';

import '../src/addr.dart';
import '../src/json.dart';
import '../src/provider.dart';
import '../src/sekreto.dart';
import '../src/support.dart';

import 'httpjson.dart';

/// The Key Vault audience an Azure token is minted for.
const String RESOURCE = 'https://vault.azure.net';

/// Azure Key Vault.
///
/// `api.token` reads secret `api-token` (dots flattened to `-`; Key Vault
/// names allow nothing else), current version. The token comes from config,
/// then a client-credentials login when tenant/clientid/clientsecret are
/// given, then the IMDS managed-identity endpoint - so on Azure's own
/// platform no credential configuration is needed.
///
/// As with GCP, the IMDS call is plain http to a link-local host by platform
/// design and carries no credential; the login and vault addresses are
/// `checkaddr`-guarded.
class Azuresecrets extends Provider {
  final String? vault;
  final String? token;
  final String? tenant;
  final String? clientid;
  final String? clientsecret;
  final String? loginaddr;
  final String? imdsaddr;
  final String? apiversion;

  // A configured token is kept forever; logged-in and IMDS tokens carry
  // expires_in and are renewed shortly before they run out.
  String? _livetoken;
  int _renewat = NEVER;

  Azuresecrets({
    this.vault,
    this.token,
    this.tenant,
    this.clientid,
    this.clientsecret,
    this.loginaddr,
    this.imdsaddr,
    this.apiversion,
  });

  bool get _hasclient =>
      null != tenant &&
      tenant!.isNotEmpty &&
      null != clientid &&
      clientid!.isNotEmpty &&
      null != clientsecret &&
      clientsecret!.isNotEmpty;

  Future<String> _login() async {
    if (null != token && token!.isNotEmpty) {
      return token!;
    }

    if (_hasclient) {
      final useloginaddr =
          first([loginaddr, 'https://login.microsoftonline.com']);
      checkaddr(useloginaddr);

      final url = '${trimslash(useloginaddr)}/${tenant!}/oauth2/v2.0/token';
      final form = 'grant_type=client_credentials'
          '&client_id=${uriescape(clientid!)}'
          '&client_secret=${uriescape(clientsecret!)}'
          '&scope=${uriescape('$RESOURCE/.default')}';

      final res = await fetchjson(
        'POST',
        url,
        headers: {'content-type': 'application/x-www-form-urlencoded'},
        body: form,
      );

      final got = res.body.dig('access_token').text;
      if (200 != res.status || null == got || got.isEmpty) {
        throw SekretoError('sekreto: azure login failed: ${res.status}');
      }

      _renewat = renewtime(res.body.dig('expires_in'));
      return got;
    }

    final imds = '${trimslash(first([imdsaddr, 'http://169.254.169.254']))}'
        '/metadata/identity/oauth2/token?api-version=2018-02-01'
        '&resource=${uriescape(RESOURCE)}';

    final res = await fetchjson('GET', imds, headers: {'Metadata': 'true'});

    final got = res.body.dig('access_token').text;
    if (200 != res.status || null == got || got.isEmpty) {
      throw SekretoError(
        'sekreto: azure: no token, no client credentials, and IMDS did not answer',
      );
    }

    // IMDS sends expires_in as a STRING, where everyone else sends a number.
    _renewat = renewtime(res.body.dig('expires_in'));
    return got;
  }

  @override
  FutureOr<String?> lookup(String name) {
    final usevault = vault ?? '';
    if (usevault.isEmpty) {
      throw SekretoError('sekreto: azure: no vault');
    }

    // Only an explicit scheme is a URL; a vault NAMED httpvault must still
    // become https://httpvault.vault.azure.net.
    final vaulturl =
        (usevault.startsWith('http://') || usevault.startsWith('https://'))
            ? usevault
            : 'https://$usevault.vault.azure.net';
    checkaddr(vaulturl);

    return _read(vaulturl, name);
  }

  Future<String?> _read(String vaulturl, String name) async {
    if (null == _livetoken || DateTime.now().millisecondsSinceEpoch >= _renewat) {
      _livetoken = await _login();
    }

    final url = '${trimslash(vaulturl)}/secrets/${flatname(name, '-')}'
        '?api-version=${first([apiversion, '7.4'])}';

    final res = await fetchjson('GET', url,
        headers: {'authorization': 'Bearer ${_livetoken ?? ''}'});

    if (404 == res.status) {
      return null;
    }
    if (200 != res.status) {
      throw SekretoError('sekreto: azure error: ${res.status}: ${bare(url)}');
    }

    return res.body.dig('value').text;
  }

  @override
  String describe() => 'azuresecrets:${vault ?? ''}';
}

/// The `azuresecrets` provider kind. Pass it in `plugins` to name it in a
/// chain.
final Definition azuresecrets = providerplugin(
  'azuresecrets',
  (spec) => Azuresecrets(
    vault: spec.vault,
    token: spec.token,
    tenant: spec.tenant,
    clientid: spec.clientid,
    clientsecret: spec.clientsecret,
    loginaddr: spec.loginaddr,
    imdsaddr: spec.imdsaddr,
    apiversion: spec.apiversion,
  ),
);

File: swift/plugins/Azuresecrets.swift
// The `azuresecrets` provider kind: Azure Key Vault.
//
// A plugin, not a built-in: it opens a socket, to the vault and to
// whichever login endpoint its credentials point at.
//
// A port of typescript/plugins/azuresecrets.ts, which is canonical.

import Foundation

import Sekreto

// A SCOPED import: plugin exports a `Json` of its own, and the only name
// this file wants from it is `Definition`.
import struct VoxgigPlugin.Definition

/// The Key Vault audience an Azure token is minted for.
let AZURERESOURCE = "https://vault.azure.net"

/// Azure Key Vault.
///
/// `api.token` reads secret `api-token` (dots flattened to `-`; Key Vault
/// names allow nothing else), current version. The token comes from
/// config, then a client-credentials login when tenant/clientid/
/// clientsecret are given, then the IMDS managed-identity endpoint - so on
/// Azure's own platform no credential configuration is needed.
///
/// As with GCP, the IMDS call is plain http to a link-local host by
/// platform design and carries no credential; the login and vault
/// addresses are `checkaddr`-guarded.
public final class AzuresecretsProvider: Provider {

  private let vault: String?
  private let token: String?
  private let tenant: String?
  private let clientid: String?
  private let clientsecret: String?
  private let loginaddr: String?
  private let imdsaddr: String?
  private let apiversion: String?

  private var livetoken: String?
  private var renewat: Double = Double.greatestFiniteMagnitude

  public init(
    vault: String? = nil,
    token: String? = nil,
    tenant: String? = nil,
    clientid: String? = nil,
    clientsecret: String? = nil,
    loginaddr: String? = nil,
    imdsaddr: String? = nil,
    apiversion: String? = nil
  ) {
    self.vault = vault
    self.token = token
    self.tenant = tenant
    self.clientid = clientid
    self.clientsecret = clientsecret
    self.loginaddr = loginaddr
    self.imdsaddr = imdsaddr
    self.apiversion = apiversion
  }

  private func login() throws -> String {
    if let use = token, !use.isEmpty { return use }

    let usetenant = tenant ?? ""
    let useclientid = clientid ?? ""
    let useclientsecret = clientsecret ?? ""

    if !usetenant.isEmpty && !useclientid.isEmpty && !useclientsecret.isEmpty {
      let useloginaddr = first(loginaddr, "https://login.microsoftonline.com")
      try checkaddr(useloginaddr)

      let url = trimslash(useloginaddr) + "/" + usetenant + "/oauth2/v2.0/token"
      let form =
        "grant_type=client_credentials&client_id=" + uriescape(useclientid)
        + "&client_secret=" + uriescape(useclientsecret)
        + "&scope=" + uriescape("\(AZURERESOURCE)/.default")

      var headers = Ordered<String>()
      headers["content-type"] = "application/x-www-form-urlencoded"

      let res = try fetchjson("POST", url, headers, form)

      let got = res.body.dig("access_token").text

      if 200 != res.status || (got?.isEmpty ?? true) {
        throw SekretoError("sekreto: azure login failed: \(res.status)")
      }

      renewat = renewtime(res.body.dig("expires_in"))

      return got!
    }

    let imds =
      trimslash(first(imdsaddr, "http://169.254.169.254"))
      + "/metadata/identity/oauth2/token?api-version=2018-02-01&resource="
      + uriescape(AZURERESOURCE)

    var headers = Ordered<String>()
    headers["Metadata"] = "true"

    let res = try fetchjson("GET", imds, headers)

    let got = res.body.dig("access_token").text

    if 200 != res.status || (got?.isEmpty ?? true) {
      throw SekretoError(
        "sekreto: azure: no token, no client credentials, and IMDS did not answer")
    }

    // IMDS sends expires_in as a STRING, unlike everybody else.
    renewat = renewtime(res.body.dig("expires_in"))

    return got!
  }

  public func lookup(_ name: String) throws -> String? {
    let usevault = vault ?? ""
    if usevault.isEmpty { throw SekretoError("sekreto: azure: no vault") }

    // Only an explicit scheme is a URL; a vault NAMED httpvault must still
    // become https://httpvault.vault.azure.net.
    let vaulturl =
      (usevault.hasPrefix("http://") || usevault.hasPrefix("https://"))
      ? usevault : "https://\(usevault).vault.azure.net"
    try checkaddr(vaulturl)

    if nil == livetoken || nowms() >= renewat {
      livetoken = try login()
    }

    let url =
      trimslash(vaulturl) + "/secrets/" + (try flatname(name, "-"))
      + "?api-version=" + first(apiversion, "7.4")

    var headers = Ordered<String>()
    headers["authorization"] = "Bearer \(livetoken ?? "")"

    let res = try fetchjson("GET", url, headers)

    if 404 == res.status { return nil }

    if 200 != res.status {
      throw SekretoError("sekreto: azure error: \(res.status): \(bare(url))")
    }

    return res.body.dig("value").text
  }

  public func describe() -> String {
    return "azuresecrets:\(vault ?? "")"
  }
}

/// The kind, as a voxgig/plugin definition.
public let azuresecrets: Definition = providerplugin("azuresecrets") { spec in
  AzuresecretsProvider(
    vault: spec.vault,
    token: spec.token,
    tenant: spec.tenant,
    clientid: spec.clientid,
    clientsecret: spec.clientsecret,
    loginaddr: spec.loginaddr,
    imdsaddr: spec.imdsaddr,
    apiversion: spec.apiversion
  )
}
